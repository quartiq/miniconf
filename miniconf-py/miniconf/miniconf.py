#!/usr/bin/python
"""
Author: Vertigo Designs, Ryan Summers
        Robert JÃ¶rdens

Description: Provides an API for controlling Miniconf devices over MQTT.
"""
import asyncio
import json
import logging
import uuid

from gmqtt import Client as MqttClient

LOGGER = logging.getLogger(__name__)


class Miniconf:
    """An asynchronous API for controlling Miniconf devices using MQTT."""

    @classmethod
    async def create(cls, prefix, broker):
        """Create a connection to the broker and a Miniconf device using it."""
        client = MqttClient(client_id='')
        await client.connect(broker)
        return cls(client, prefix)

    def __init__(self, client, prefix):
        """Constructor.

        Args:
            client: A connected MQTT5 client.
            prefix: The MQTT toptic prefix of the device to control.
        """
        self.request_id = 0
        self.client = client
        self.prefix = prefix
        self.inflight = {}
        self.client.on_message = self._handle_response
        self.response_topic = f'{prefix}/response/{uuid.uuid1().hex}'
        self.client.subscribe(self.response_topic)

    def _handle_response(self, _client, topic, payload, _qos, properties):
        """Callback function for when messages are received over MQTT.

        Args:
            _client: The MQTT client.
            topic: The topic that the message was received on.
            payload: The payload of the message.
            _qos: The quality-of-service level of the received packet
            properties: A dictionary of properties associated with the message.
        """
        if topic == self.response_topic:
            # Extract request_id corrleation data from the properties
            request_id = int.from_bytes(
                properties['correlation_data'][0], 'big')

            self.inflight[request_id].set_result(json.loads(payload))
            del self.inflight[request_id]
        else:
            LOGGER.warning('Unexpected message on "%s"', topic)

    async def command(self, path, value, retain=True):
        """Write the provided data to the specified path.

        Args:
            path: The path to write the message to.
            value: The value to write to the path.
            retain: Retain the MQTT message changing the setting
                by the broker.

        Returns:
            The response to the command as a dictionary.
        """
        topic = f'{self.prefix}/settings/{path}'

        fut = asyncio.get_running_loop().create_future()

        # Assign unique correlation data for response dispatch
        assert self.request_id not in self.inflight
        self.inflight[self.request_id] = fut
        correlation_data = self.request_id.to_bytes(4, 'big')
        self.request_id += 1

        payload = json.dumps(value)
        LOGGER.info('Sending "%s" to "%s"', value, topic)

        self.client.publish(
            topic, payload=payload, qos=0, retain=retain,
            response_topic=self.response_topic,
            correlation_data=correlation_data)

        return await fut
