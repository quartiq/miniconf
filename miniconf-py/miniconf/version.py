#!/usr/bin/python3
""" Miniconf version file. """

# The semantic version of Miniconf.
__version__ = '0.1.0'
