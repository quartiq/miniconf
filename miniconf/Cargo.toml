[package]
name = "miniconf"
# Sync all crate versions and the py client
version = "0.19.0"
authors = [
    "James Irwin <irwineffect@gmail.com>",
    "Ryan Summers <ryan.summers@vertigo-designs.com>",
    "Robert JÃ¶rdens <rj@quartiq.de>",
]
edition = "2021"
license = "MIT"
description = "Serialize/deserialize/access reflection for trees"
repository = "https://github.com/quartiq/miniconf"
keywords = ["config", "serde", "no_std", "reflection", "graph"]
categories = ["embedded", "config", "data-structures", "parsing"]

[dependencies]
serde = { version = "1.0.120", default-features = false }
miniconf_derive = { path = "../miniconf_derive", version = "0.19.0", optional = true }
itoa = "1.0.4"
serde-json-core = { version = "0.6.0", optional = true }
postcard = { version = "1.0.8", optional = true }
thiserror = { version = "2", default-features = false }
serde-reflection = { version = "0.5", git = "https://github.com/quartiq/serde-reflection.git", branch = "pub-ser-de", optional = true }
once_cell = { version = "1.21.3", default-features = false, optional = true }
schemars = { version = "1.0.4", optional = true, default-features = false }
serde_json = { version = "1.0.133", optional = true }

[features]
default = ["derive", "trace"]
alloc = []
std = ["alloc"]
json-core = ["dep:serde-json-core"]
postcard = ["dep:postcard"]
derive = ["dep:miniconf_derive", "serde/derive"]
trace = ["std", "dep:serde-reflection", "dep:once_cell"]
schema = ["dep:schemars", "dep:serde_json", "trace"]

[package.metadata.docs.rs]
all-features = true

[dev-dependencies]
anyhow = "1.0.86"
postcard = { version = "1.0.8", features = ["use-std"] }
crosstrait = { version = "0.1", default-features = false }
embedded-io = "0.6.1"
embedded-io-async = "0.6.1"
embedded-io-adapters = { version = "0.6.1", features = ["tokio-1"] }
heapless = "0.8.0"
yafnv = "3.0.0"
tokio = { version = "1.38.0", features = ["io-std", "rt", "macros"] }
strum = { version = "0.27.1", features = ["derive"] }
trybuild = { version = "1.0.99", features = ["diff"] }
serde_json = "1.0.133"
jsonschema = { version = "0.32.1", default-features = false }

[[test]]
name = "arrays"
required-features = ["json-core", "derive"]

[[test]]
name = "basic"
required-features = ["json-core", "derive"]

[[test]]
name = "generics"
required-features = ["json-core", "derive"]

[[test]]
name = "iter"
required-features = ["json-core", "derive"]

[[test]]
name = "option"
required-features = ["json-core", "derive"]

[[test]]
name = "packed"
required-features = ["json-core", "derive"]

[[test]]
name = "skipped"
required-features = ["derive"]

[[test]]
name = "structs"
required-features = ["json-core", "derive"]

[[test]]
name = "enum"
required-features = ["json-core", "derive"]

[[test]]
name = "validate"
required-features = ["json-core", "derive"]

[[test]]
name = "flatten"
required-features = ["json-core", "derive"]

[[test]]
name = "compiletest"
required-features = ["derive"]

[[example]]
name = "common"
crate-type = ["lib"]
required-features = ["derive"]

[[example]]
name = "cli"
required-features = ["json-core", "derive"]

[[example]]
name = "scpi"
required-features = ["json-core", "derive"]

[[example]]
name = "trace"
required-features = ["derive", "schema"]
