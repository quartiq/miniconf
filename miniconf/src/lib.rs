#![cfg_attr(not(any(test, feature = "std")), no_std)]
#![cfg_attr(all(feature = "derive", feature = "json-core"), doc = include_str!("../README.md"))]
#![cfg_attr(not(all(feature = "derive", feature = "json-core")), doc = "Miniconf")]
#![deny(rust_2018_compatibility)]
#![deny(rust_2018_idioms)]
// #![warn(missing_docs)]
#![forbid(unsafe_code)]

mod error;
pub use error::*;
mod key;
pub use key::*;
mod node;
pub use node::*;
mod packed;
pub use packed::*;
mod jsonpath;
pub use jsonpath::*;
mod tree;
pub use tree::*;
mod iter;
pub use iter::*;
mod impls;
mod leaf;
pub use leaf::*;
mod walk;
pub use walk::*;

#[cfg(feature = "derive")]
pub use miniconf_derive::*;

#[cfg(feature = "json-core")]
pub mod json;

#[cfg(feature = "postcard")]
pub mod postcard;

#[cfg(feature = "alloc")]
extern crate alloc;

#[cfg(feature = "trace")]
pub mod trace;

// re-export for proc-macro
#[doc(hidden)]
pub use serde::{Deserialize, Deserializer, Serialize, Serializer};
