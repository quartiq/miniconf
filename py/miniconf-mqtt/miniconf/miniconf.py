#!/usr/bin/python
"""
Author: Vertigo Designs, Ryan Summers
        Robert JÃ¶rdens

Description: Provides an API for controlling Miniconf devices over MQTT.
"""
import asyncio
import json
import logging
import uuid

from aiomqtt import Client, Message, MqttError
from paho.mqtt.properties import Properties, PacketTypes
import paho.mqtt

MQTTv5 = paho.mqtt.enums.MQTTProtocolVersion.MQTTv5

LOGGER = logging.getLogger(__name__)


class MiniconfException(Exception):
    """Generic exceptions generated by Miniconf."""


class Miniconf:
    """An asynchronous API for controlling Miniconf devices using MQTT."""

    def __init__(self, client: Client, prefix: str):
        """Constructor.

        Args:
            client: A connected MQTT5 client.
            prefix: The MQTT toptic prefix of the device to control.
        """
        self.client = client
        self.prefix = prefix
        # A dispatcher is required since mqtt does not guarantee in-order processing
        # across topics (within a topic processing is mostly in-order).
        # Responses to requests on different topics may arrive out-of-order.
        self._inflight = {}
        self.response_topic = f"{prefix}/response"
        self.listener = asyncio.create_task(self._listen())
        self.subscribed = asyncio.Event()

    async def _listen(self):
        await self.client.subscribe(self.response_topic)
        LOGGER.info(f"Subscribed to {self.response_topic}")
        self.subscribed.set()
        try:
            async for message in self.client.messages:
                self._dispatch(message)
        except (asyncio.CancelledError, MqttError):
            pass
        finally:
            try:
                await self.client.unsubscribe(self.response_topic)
                LOGGER.info(f"Unsubscribed from {self.response_topic}")
            except MqttError:
                pass

    def _dispatch(self, message: Message):
        if message.topic.value != self.response_topic:
            LOGGER.warning(
                "Discarding message with unexpected topic: %s", message.topic.value
            )
            return

        try:
            properties = message.properties.json()
        except AttributeError:
            properties = {}
        # lazy formatting
        LOGGER.debug("Received %s: %s [%s]", message.topic, message.payload, properties)

        try:
            response_id = bytes.fromhex(properties["CorrelationData"])
        except KeyError:
            LOGGER.info("Discarding message without CorrelationData")
            return
        try:
            fut, ret = self._inflight[response_id]
        except KeyError:
            LOGGER.info(
                f"Discarding message with unexpected CorrelationData: {response_id}"
            )
            return

        try:
            code = dict(properties["UserProperty"])["code"]
        except KeyError:
            LOGGER.warning("Discarding message without response code user property")
            return

        response = message.payload.decode("utf-8")
        if code == "Continue":
            ret.append(response)
            return

        if code == "Ok":
            if response:
                ret.append(response)
            fut.set_result(ret)
        else:
            fut.set_exception(
                MiniconfException(f"Received code: {code}, Message: {response}")
            )
        del self._inflight[response_id]

    async def _do(self, topic: str, *, response=True, **kwargs):
        await self.subscribed.wait()

        request_id = uuid.uuid1().bytes
        fut = asyncio.get_event_loop().create_future()
        assert request_id not in self._inflight
        self._inflight[request_id] = fut, []

        props = Properties(PacketTypes.PUBLISH)
        if response:
            props.ResponseTopic = self.response_topic
        props.CorrelationData = request_id
        LOGGER.info(f"Publishing {topic}: {kwargs['payload']}, [{props}]")
        await self.client.publish(
            topic,
            properties=props,
            **kwargs,
        )
        if response:
            return await fut

    async def set(self, path, value, retain=False):
        """Write the provided data to the specified path.

        Args:
            path: The path to set.
            value: The value to set.
            retain: Retain the the setting on the broker.
        """
        ret = await self._do(
            topic=f"{self.prefix}/settings{path}",
            payload=json.dumps(value, separators=(",", ":")),
            retain=retain,
        )
        assert len(ret) == 1, ret
        return ret[0]

    async def list_paths(self, root=""):
        """Get a list of all the paths below a given root.

        Args:
            root: Path to the root node to list.
        """
        return await self._do(topic=f"{self.prefix}/settings{root}", payload="")

    async def dump(self, root=""):
        """Dump all the paths at or below a given root into the settings namespace.

        Note that the target Miniconf client may be unable to
        responde to messages when a multipart operation (list or dump) is in progress.
        This method does not wait for the completion of the dump.

        Args:
            root: Path to the root node to dump.
        """
        await self._do(
            topic=f"{self.prefix}/settings{root}", payload="", response=False
        )

    async def get(self, path):
        """Get the specific value of a given path.

        Args:
            path: The path to get.
        """
        ret = await self._do(topic=f"{self.prefix}/settings{path}", payload="")
        assert len(ret) == 1, ret
        return ret[0]

    async def clear(self, path):
        """Clear retained value from a path.

        Args:
            path: The path to get.
        """
        ret = await self._do(f"{self.prefix}/settings{path}", payload="", retain=True)
        assert len(ret) == 1, ret
        return ret[0]
