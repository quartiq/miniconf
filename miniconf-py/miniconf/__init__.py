#!/usr/bin/python3
from .miniconf import Miniconf
from .version import __version__
