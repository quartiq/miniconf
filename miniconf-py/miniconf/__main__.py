#!/usr/bin/python3
"""
Author: Ryan Summers, Robert JÃ¶rdens

Description: Command-line utility to program run-time settings utilize Miniconf.
"""

import sys
import argparse
from .miniconf import Miniconf


def main():
    """ Main program entry point. """
    parser = argparse.ArgumentParser(
        description='Miniconf command line interface.',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog='''Examples:
%(prog)s dt/sinara/dual-iir/00-11-22-33-aa-bb stream_target=\
'{"ip": [192, 168, 0, 1], "port": 1000}'
''')
    parser.add_argument('-v', '--verbose', action='count', default=0,
                        help='Increase logging verbosity')
    parser.add_argument('--broker', '-b', default='mqtt', type=str,
                        help='The MQTT broker address')
    parser.add_argument('--no-retain', '-n', default=False,
                        action='store_true',
                        help='Do not retain the affected settings')
    parser.add_argument('prefix', type=str,
                        help='The MQTT topic prefix of the target')
    parser.add_argument('settings', metavar="PATH=VALUE", nargs='+',
                        help='JSON encoded values for settings path keys.')

    args = parser.parse_args()

    logging.basicConfig(
        format='%(asctime)s [%(levelname)s] %(name)s: %(message)s',
        level=logging.WARN - 10*args.verbose)

    loop = asyncio.get_event_loop()

    async def configure_settings():
        interface = await Miniconf.create(args.prefix, args.broker)
        for setting in args.settings:
            path, value = setting.split("=", 1)
            response = await interface.command(path, json.loads(value),
                                               not args.no_retain)
            print(f'{path}: {response}')
            if response['code'] != 0:
                return response['code']
        return 0

    sys.exit(loop.run_until_complete(configure_settings()))


if __name__ == '__main__':
    main()
