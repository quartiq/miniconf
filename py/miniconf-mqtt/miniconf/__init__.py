#!/usr/bin/python3
""" Root Miniconf module file. """
import asyncio
import json
import logging
import time

from typing import Set

from gmqtt import Client as MqttClient

from .miniconf import Miniconf
from .version import __version__

async def discover(
        broker: str,
        prefix_filter: str,
        discovery_timeout: float = 0.1,
    ) -> Set[str]:
    """ Get a list of available Miniconf devices.

    Args:
        * `broker` - The broker to search for clients on.
        * `prefix_filter` - An MQTT-specific topic filter for device prefixes. Note that this will
          be appended to with the default status topic name `/alive`.
        * `discovery_timeout` - The duration to search for clients in seconds.

    Returns:
        A set of discovered client prefixes that match the provided filter.
    """
    discovered_devices = set()

    suffix = '/alive'

    def handle_message(_client, topic, payload, _qos, _properties):
        logging.debug('Got message from %s: %s', topic, payload)

        if json.loads(payload):
            discovered_devices.add(topic[:-len(suffix)])

    client = MqttClient(client_id='')
    client.on_message = handle_message

    await client.connect(broker)

    client.subscribe(f'{prefix_filter}{suffix}')

    await asyncio.sleep(discovery_timeout)

    return discovered_devices
