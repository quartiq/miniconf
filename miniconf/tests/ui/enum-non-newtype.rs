use miniconf::Tree;

#[derive(Tree)]
pub enum E1 {
    A(i32, i32),
}

fn main() {}
