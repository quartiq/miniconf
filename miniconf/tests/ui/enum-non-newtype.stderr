error: Only newtype (single field tuple) enum variants are supported.
 --> tests/ui/enum-non-newtype.rs:5:5
  |
5 |     A(i32, i32),
  |     ^

error: Unsupported shape `named fields`. Expected unnamed fields or no fields.
 --> tests/ui/enum-non-newtype.rs:8:10
  |
8 | #[derive(Tree)]
  |          ^^^^
  |
  = note: this error originates in the derive macro `Tree` (in Nightly builds, run with -Z macro-backtrace for more info)
